std/numeric.d

```diff
@@ -1400,6 +1400,270 @@ unittest
+        R q = 0;
+        R r = 0;
+
+        // fit parabola
+        if (fabs(e) > tolerance)
+        {
+            immutable  xw =  x -  w;
+            immutable fxw = fx - fw;
+            immutable  xv =  x -  v;
+            immutable fxv = fx - fv;
+            immutable xwfxv = xw * fxv;
+            immutable xvfxw = xv * fxw;
+            p = xv * xvfxw - xw * xwfxv;
+            q = (xvfxw - xwfxv) * 2;
+            if (q > 0)
+                p = -p;
+            else
+                q = -q;
+            r = e;
+            e = d;
+        }
+
+        T u;
+        // a parabolic-interpolation step
+        if (fabs(p) < fabs(q * r / 2) && p > q * (a - x) && p < q * (b - x))
+        {
+            d = p / q;
+            u = x + d;
+            // f must not be evaluated too close to a or b
+            if (u - a < t2 || b - u < t2)
+                d = x < m ? tolerance : -tolerance;
+        }
+        // a golden-section step
+        else
+        {
+            e = (x < m ? b : a) - x;
+            d = c * e;
+        }
+
+        // f must not be evaluated too close to x
+        u = x + (fabs(d) >= tolerance ? d : d > 0 ? tolerance : -tolerance);
+        immutable fu = f(u);
+        if(isNaN(fu) || fu == -T.infinity)
+            return typeof(return)(u, fu, T.init);
+
+        //  update  a, b, v, w, and x
+        if (fu <= fx)
+        {
+            u < x ? b : a = x;
+            v = w; fv = fw;
+            w = x; fw = fx;
+            x = u; fx = fu;
+        }
+        else
+        {
+            u < x ? a : b = u;
+            if (fu <= fw || w == x)
+            {
+                v = w; fv = fw;
+                w = u; fw = fu;
+            }
+            else if (fu <= fv || v == x || v == w)
+            { // do not remove this braces
+                v = u; fv = fu;
+            }
+        }
+    }
+    return typeof(return)(x, fx, tolerance * 3);
+}
+
+///
+unittest
+{
+    auto ret = findLocalMin((double x) => (x-4)^^2, -1e9, 1e9);
+    assert(ret.x.approxEqual(4.0));
+    assert(ret.y.approxEqual(0.0));
+}
+
+unittest
+{
+    int i;
+    double f(double x)
+    {
+        i++;
+        return fabs(x-1);
+    }
+
+    //slow
+    auto minD = findLocalMin(&f, -double.max/2, double.max/2, double.min_normal, 2*double.epsilon);
+    with(minD)
+    {
+        assert(approxEqual(x, 1));
+        assert(approxEqual(y, 0));
+        assert(error <= 10 * double.epsilon);
+        //assert(minD.error == 0);
+    }
+}
+
+unittest
+{
+    auto ret = findLocalMin((double x) => double.init, 0.0, 1.0, double.min_normal, 2*double.epsilon);
+    assert(!ret.x.isNaN);
+    assert(ret.y.isNaN);
+    assert(ret.error.isNaN);
+}
+
+unittest
+{
+    auto ret = findLocalMin((double x) => log(x), 0.0, 1.0, double.min_normal, 2*double.epsilon);
+    assert(ret.error < 3.00001 * ((2*double.epsilon)*fabs(ret.x)+ double.min_normal));
+    assert(ret.x >= 0 && ret.x <= ret.error);
+}
+
+
+unittest
+{
+    size_t i;
```

Comment:
What is the purpose of `i` here? (I can see it counts function evaluations, but it's not used for anything afterwards.)